File: sections/42_advanced_data_structures_mixed/problems/f_dynamic_path_maximum/solution.py
import sys
def main():
    d=sys.stdin.buffer.read().split()
    if not d:return
    n,q=map(int,d[:2]);a=list(map(int,d[2:2+n]));g=[[]for _ in range(n)];j=2+n
    for _ in range(n-1):u,v=int(d[j])-1,int(d[j+1])-1;j+=2;g[u].append(v);g[v].append(u)
    par=[-1]*n;dep=[0]*n;order=[0]
    for u in order:
        for v in g[u]:
            if v!=par[u]:par[v]=u;dep[v]=dep[u]+1;order.append(v)
    size=[1]*n;heavy=[-1]*n
    for u in reversed(order[1:]):
        size[par[u]]+=size[u]
        if heavy[par[u]]<0 or size[u]>size[heavy[par[u]]]:heavy[par[u]]=u
    head=[0]*n;pos=[0]*n;timer=0;todo=[(0,0)]
    while todo:
        u,h=todo.pop()
        while u!=-1:
            head[u]=h;pos[u]=timer;timer+=1
            for v in g[u]:
                if v!=par[u] and v!=heavy[u]:todo.append((v,v))
            u=heavy[u]
    z=1
    while z<n:z*=2
    tree=[-10**30]*(2*z)
    for u in range(n):tree[z+pos[u]]=a[u]
    for i in range(z-1,0,-1):tree[i]=max(tree[2*i],tree[2*i+1])
    def rq(l,r):
        ans=-10**30;l+=z;r+=z+1
        while l<r:
            if l&1:ans=max(ans,tree[l]);l+=1
            if r&1:r-=1;ans=max(ans,tree[r])
            l//=2;r//=2
        return ans
    out=[]
    for _ in range(q):
        c=d[j];u=int(d[j+1])-1;v=int(d[j+2]);j+=3
        if c==b'U':
            x=z+pos[u];tree[x]=v;x//=2
            while x:tree[x]=max(tree[2*x],tree[2*x+1]);x//=2
        else:
            v-=1;ans=-10**30
            while head[u]!=head[v]:
                if dep[head[u]]<dep[head[v]]:u,v=v,u
                ans=max(ans,rq(pos[head[u]],pos[u]));u=par[head[u]]
            if dep[u]>dep[v]:u,v=v,u
            out.append(str(max(ans,rq(pos[u],pos[v]))))
    print('\n'.join(out))
if __name__=="__main__":main()

File: sections/42_advanced_data_structures_mixed/problems/f_dynamic_path_maximum/solution.cpp
#include <bits/stdc++.h>
using namespace std;
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q))
    return 0;
  vector<int> a(n), p(n, -1), dep(n), sz(n, 1), heavy(n, -1), head(n), pos(n),
      ord = {0};
  for (int &x : a)
    cin >> x;
  vector<vector<int>> g(n);
  for (int i = 1, u, v; i < n; i++) {
    cin >> u >> v;
    --u;
    --v;
    g[u].push_back(v);
    g[v].push_back(u);
  }
  for (int i = 0; i < (int)ord.size(); i++) {
    int u = ord[i];
    for (int v : g[u])
      if (v != p[u])
        p[v] = u, dep[v] = dep[u] + 1, ord.push_back(v);
  }
  for (int z = n - 1; z; z--) {
    int u = ord[z];
    sz[p[u]] += sz[u];
    if (heavy[p[u]] < 0 || sz[u] > sz[heavy[p[u]]])
      heavy[p[u]] = u;
  }
  int timer = 0;
  vector<pair<int, int>> todo = {{0, 0}};
  while (!todo.empty()) {
    auto [u, h] = todo.back();
    todo.pop_back();
    for (int x = u; x != -1; x = heavy[x]) {
      head[x] = h;
      pos[x] = timer++;
      for (int v : g[x])
        if (v != p[x] && v != heavy[x])
          todo.push_back({v, v});
    }
  }
  int z = 1;
  while (z < n)
    z *= 2;
  vector<int> t(2 * z, INT_MIN);
  for (int u = 0; u < n; u++)
    t[z + pos[u]] = a[u];
  for (int i = z - 1; i; i--)
    t[i] = max(t[2 * i], t[2 * i + 1]);
  auto query = [&](int l, int r) {
    int ans = INT_MIN;
    for (l += z, r += z + 1; l < r; l /= 2, r /= 2) {
      if (l & 1)
        ans = max(ans, t[l++]);
      if (r & 1)
        ans = max(ans, t[--r]);
    }
    return ans;
  };
  while (q--) {
    char c;
    int u, v;
    cin >> c >> u >> v;
    --u;
    if (c == 'U') {
      int x = z + pos[u];
      t[x] = v;
      for (x /= 2; x; x /= 2)
        t[x] = max(t[2 * x], t[2 * x + 1]);
    } else {
      --v;
      int ans = INT_MIN;
      while (head[u] != head[v]) {
        if (dep[head[u]] < dep[head[v]])
          swap(u, v);
        ans = max(ans, query(pos[head[u]], pos[u]));
        u = p[head[u]];
      }
      if (dep[u] > dep[v])
        swap(u, v);
      cout << max(ans, query(pos[u], pos[v])) << '\n';
    }
  }
}
